Sources/PostgresNIO/Data/PostgresData+Range.swift

```diff
@@ -0,0 +1,244 @@
+import NIOCore
+
+// MARK: PostgresRangeBound
+
+public protocol PostgresRangeBound: PostgresDataConvertible, Comparable {
+    /// the Postgres range type associated with this bound
+    static var rangeType: PostgresDataType { get }
+    
+    /// If a Postgres range type has a well-defined step,
+    /// Postgres automatically converts it to a canonical form.
+    /// Types such as `int4range` get converted to upper-bound-exclusive.
+    /// This method is needed when converting an upper bound to inclusive.
+    var stepDown: (() -> Self)? { get }
+}
+
+extension Int32: PostgresRangeBound {
+    public static var rangeType: PostgresDataType { return .int4Range }
+
+    public var stepDown: (() -> Int32)? {
+        { self - 1 }
+    }
+}
+
+extension Int64: PostgresRangeBound {
+    public static var rangeType: PostgresDataType { return .int8Range }
+
+    public var stepDown: (() -> Int64)? {
+        { self - 1 }
+    }
+}
+
+// MARK: PostgresRange
+
+public struct PostgresRange<B: PostgresRangeBound>: CustomStringConvertible {
```

Comment:
Do we need to make this type public?

Reply:
If I remove `public`, I see build errors such as `generic struct 'PostgresRange' is internal and cannot be referenced from an '@inlinable' function` on this line: https://github.com/vapor/postgres-nio/pull/330/files/61b4d5939e43d88ce92406c07645427dd561ac4e#diff-5c0292b699e71becd70812877bc88e53d4e1c18d061b195dacfcdfbb0d8f8faaR17. I could try removing `@inlinable` but the other existing conformances to `PostgresDecodable` and `PostgresEncodable` use it so I figured I should leave it.

Reply:
Another solution would be to add `@usableFromInline` in a bunch of places, such as:
```
@usableFromInline
struct PostgresRange<B> {
    @usableFromInline let lowerBound: B?
    @usableFromInline let upperBound: B?
    @usableFromInline let isLowerBoundInclusive: Bool
    @usableFromInline let isUpperBoundInclusive: Bool

    init(
        lowerBound: B?,
        upperBound: B?,
        isLowerBoundInclusive: Bool,
        isUpperBoundInclusive: Bool
    ) {
        self.lowerBound = lowerBound
        self.upperBound = upperBound
        self.isLowerBoundInclusive = isLowerBoundInclusive
        self.isUpperBoundInclusive = isUpperBoundInclusive
    }
}
```

Reply:
yes please let's go for the second option! I really, really only want to add things to the public API that we need.